### PR TITLE
added a csv reader with parsing to a matrix

### LINKED ISSUE #1: read csv data and parse to a matrix

### PR DESCRIPTION
**Pull request by nickBes**
as per #1, closes #1

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
+// we have only one form so we can use querySelector
+let form = document.querySelector('form')
```
Comment (MaderNoob) [inline index.js:2]: use `const`

File: index.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+let matrix;
```
Comment (MaderNoob) [inline index.js:3]: Add some documentation explaining what this is.

File: index.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+form.onchange = async () => {
+    const data = new FormData(form)
+    // if form doesn't have an input of name file stop
+    if (!data.has('file')) return
```
Comment (MaderNoob) [inline index.js:8]: What about some indication to the user that he needs to upload a file.

File: index.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    let file = data.get('file')
```
Comment (MaderNoob) [inline index.js:9]: use `const`

File: index.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    if (file instanceof File) {
+        const text = await file.text()
+        matrix = parseTextCSV(text)
+        console.log(matrix)
+    }
+}
+
+function parseTextCSV (text) {
+    if (typeof text === "string") {
```
Comment (MaderNoob) [inline index.js:18]: When would it not be a string?

File: index.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        let parsedRows = text.split('\n') // seperate rows
```
Comment (MaderNoob) [inline index.js:19]: Some files will use `'\r\n'` as line endings. You could maybe use regex to allow both `'\n'` and `'\r\n'`:

I think something like this should work:
```js
/(\n)|(\r\n)/
```

File: index.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+                        .map(str => { // parse data into a matrix
```
Comment (MaderNoob) [inline index.js:20]: call it `row`

File: index.js [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+                            // this might set numbers as NaN when string is given
+                            let [x, y] = str.split(',').map(val => parseFloat(val))
```
Comment (MaderNoob) [inline index.js:22]: What happens when `str` contains no comma, or more than 2 commas?

File: index.js [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+                            // will return 0 for invalid values
+                            if (isInvalidNum(x)) {
+                                return 0
```
Comment (MaderNoob) [inline index.js:25]: Maybe show some error message instead?

File: index.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+function parseTextCSV (text) {
+    if (typeof text === "string") {
+        let parsedRows = text.split('\n') // seperate rows
+                        .map(str => { // parse data into a matrix
+                            // this might set numbers as NaN when string is given
+                            let [x, y] = str.split(',').map(val => parseFloat(val))
+                            // will return 0 for invalid values
+                            if (isInvalidNum(x)) {
```
Comment (MaderNoob) [inline index.js:24]: Why are you only checking `x`?

File: index.js [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+                                return 0
+                            }
+
+                            return [x, y]
+                        })
+                        .filter(val => val) // filter 0 values
```
Comment (MaderNoob) [inline index.js:30]: How does this filter 0 values if `val` is an array of `[x, y]`?

**Review by MaderNoob — CHANGES_REQUESTED** [review]
Looks good, added a few questions and things to fix. Most comments are related to advanced topics which we'll probably only want to implement in a future. You should decide what is relevant to this PR, and what requires its own issue.

File: index.js (hunk 3 above)
Comment (nickBes) [reply index.js:8]: this is onchange, so to get a message he has to upload a file

File: index.js (hunk 5 above)
Comment (nickBes) [reply index.js:18]: I wanted type annotations

File: index.js (hunk 9 above)
Comment (nickBes) [reply index.js:25]: see #3

File: index.js (hunk 8 above)
Comment (nickBes) [reply index.js:22]: also #3

File: index.js (hunk 10 above)
Comment (nickBes) [reply index.js:24]: holes

**Pull request merged**